InputEvent as_text now returns readable string. Added to_string for debug strings

EricEzaM opened the pull request:
InputEvent changed as_text() to return a display-friendly and human readable string, to be used for UI display. to_string() can be used to get a non-display friendly version which has more information for debugging. Before, some as_text() methods returned a nice readable string, and some displayed more debug-style information.

Based on Calinou comment https://github.com/godotengine/godot/issues/38326#issuecomment-621842765

This is the first commit of a larger series based on #42600.

EricEzaM committed:
InputEvent as_text now returns readable string. Added to_string for debug strings

Made InputEvent as_text() return a readable and presentable string. Added to_string() overrides for each which returns a 'debug-friendly' version which is not as presentable but provides more information and in a more structured fashion. Use as_text() for UI display scenarions and to_string() for debug cases

akien-mga commented:
This will likely conflict with #43591 (or the other way around, depending on which is merged first).

groud reviewed (approved):
Looks good.

madmiraal commented on core/input/input_event.cpp:
```diff
@@ -762,7 +834,40 @@ bool InputEventJoypadMotion::action_match(const Ref<InputEvent> &p_event, bool *
+	TTRC("Joystick 4 Up"),
+	TTRC("Joystick 4 Down"),
+};
+
```
These shouldn't be defined twice:
https://github.com/godotengine/godot/blob/c9bd11a479099bf9d0e9d3cb61f9bf362e76344e/editor/input_map_editor.cpp#L56-L77
There should be consistency between the Editor Shortcut descriptions and the Input Mapping descriptions. And, it probably makes sense to bring the descriptions for the Input Mapping here as part of this PR.
However, as [mentioned](https://github.com/godotengine/godot/pull/43660#issuecomment-730221162) by @akien-mga, these are in the process of being updated in #43591; so let's make sure they make sense in both contexts.

madmiraal reviewed (changes requested):
See below for specifics, but this PR includes a duplication of descriptions for joypad buttons and axes that needs to be addressed.

madmiraal commented on core/input/input_event.cpp:
```diff
@@ -835,7 +940,39 @@ bool InputEventJoypadButton::shortcut_match(const Ref<InputEvent> &p_event) cons
+	TTRC("D-Pad Left"),
+	TTRC("D-Pad Right")
+};
+
```
Ditto:
https://github.com/godotengine/godot/blob/c9bd11a479099bf9d0e9d3cb61f9bf362e76344e/editor/input_map_editor.cpp#L38-L54

EricEzaM commented:
@madmiraal ah yep in the original pr of #42600  I completely remove the input map editor, as I make it generic (action_map_editor.cpp), so in that pr this issue does not exist. Here it seems I will have to edit the input map editor to use as_text rather than constructing its own strings everywhere.

Seeing input map editor construct the representation of the events on its own is what made me want to do this change anyway, since the text representation of an event should be consistent across the engine.

EricEzaM commented:
@madmiraal @akien-mga Ok, finished the above reviews up. Also added some text to the descriptions as per https://github.com/godotengine/godot/pull/43591/files#r52860139, and also rebased with newest master, which includes #43591

madmiraal commented on core/input/input_event.cpp:
```diff
@@ -796,7 +868,40 @@ bool InputEventJoypadMotion::action_match(const Ref<InputEvent> &p_event, bool *
+	}
+
+	return vformat(RTR("Joypad Motion on Axis %s%s with Value %s"), itos(axis), description, String(Variant(axis_value)));
+}
```
I think there is something wrong here. It looks like you're missing the `+` or `-` after the axis number and before the description in parentheses and instead are adding "with Value ±1" to the end. Is this intentional? This is what the Input Map Editor currently presents:

![Axis Selection](https://user-images.githubusercontent.com/9253928/99678818-237e0880-2a73-11eb-83ec-334b4c45e6ed.png)

EricEzaM replied:
Yeah it looks wrong:

![image](https://user-images.githubusercontent.com/41730826/99960243-26ad2780-2dd8-11eb-8c95-4cd98db7a17b.png)

What about "Joypad Motion on Axis 0, Value -1 (Left Stick Left, Joystick 0 Left)", i.e. putting the value before the description. I think I actually meant to do this, as how I currently have it in the image above doesn't make much sense.

madmiraal replied:
Perhaps there does need to be a separation between the `as_text()` for the event and the 20 direction descriptions used by Input Map Editor.

EricEzaM replied:
It would be nice if it could be avoided but maybe it can't. To keep the descriptions in one place though, a method to get the axis/button descriptions would need to be introduced (i.e. `get_axis_description(int axis)`)

Is there a particular issue you have with `Joypad Motion on Axis 0, Value -1 (Left Stick Left, Joystick 0 Left)`? This will also likely come along with a different event editor (shown below) which was a part of #42600 which I have now split up into multiple prs.

![new_dialog_demo](https://user-images.githubusercontent.com/41730826/99963449-8b1eb580-2ddd-11eb-8dc2-3bafdc718a66.gif)

madmiraal replied:
> Is there a particular issue you have with Joypad Motion on Axis 0, Value -1 (Left Stick Left, Joystick 0 Left)?

I just think there is a difference between the event description, which is axis specific and the Input Map description, which is half axis specific:
Event description: `Joypad Motion on Axis 0, (Left Stick X-Axis, Joystick 0 X-Axis) with Value -0.3`
Input Map description: `Joypad Axis 0 - (Left Stick Left, Joystick 0 Left)`

EricEzaM replied:
Ok, I added a method called `get_axis_description()` so that you can get the axis description for a given `InputEventJoypadMotion`. This means the list of descriptions is still in once place, but the description for an event can still be accessed from anywhere. Thus, the text can still be constructed in `input_map_editor.cpp` but it is all on one line as opposed to previously many lines used to construct the description.

`vformat(TTR("Axis %s%s (%s)"), jm->get_axis(), jm->get_axis_value() < 0 ? "-" : "+", jm->get_axis_description());`

madmiraal replied:
I still think the differences between the event descriptions (10 axis specific), and the Input Map descriptions (20 half-axis specific) are enough to justify two different lists of descriptions for the two different needs:

**Event descriptions:**
- Joypad **Motion on** Axis 0, (Left Stick **X-Axis**, Joystick 0 **X-Axis**) **with Value -0.3**
- Joypad **Motion on** Axis 0, (Left Stick **X-Axis**, Joystick 0 **X-Axis**) **with Value 0.3**
- Joypad **Motion on** Axis 1, (Left Stick **Y-Axis**, Joystick 0 **Y-Axis**) **with Value -0.3**
- Joypad **Motion on** Axis 1, (Left Stick **Y-Axis**, Joystick 0 **Y-Axis**) **with Value 0.3**

**Input Map descriptions:**
- Joypad Axis 0 **–** (Left Stick **Left**, Joystick 0 **Left**)
- Joypad Axis 0 **+** (Left Stick **Right**, Joystick 0 **Right**)
- Joypad Axis 1 **–** (Left Stick **Up**, Joystick 0 **Up**)
- Joypad Axis 1 **+** (Left Stick **Down**, Joystick 0 **Down**)

I've highlighted the differences above for one joystick (two axes and the four directions).

EricEzaM replied:
So do you propose a method specific for InputEventJoypadMotion? In addition to `as_text()`, something like `input_map_description()` (bad example but you know what I mean)?

akien-mga replied:
If it's specific to the input map, it should only be in the input map editor code IMO and not an extra method in InputEvents.

EricEzaM replied:
Yep, makes sense

EricEzaM replied:
Ok, all done. Changed `as_text()` to contain 10 options for InputEventJoyMotion, and added joy axis descriptions to Input Map Editor such that the description meets the above examples by @madmiraal.

See lines 358-390 of input_map_editor.cpp for the main changes in this iteration

EricEzaM commented on core/input/input_event.cpp:
```diff
@@ -88,6 +88,10 @@ String InputEvent::as_text() const {
+String InputEvent::to_string() {
+	return String();
+}
+
```
I dont think it can be abstract since to_string() is an override so should do something.

Maybe just "Input Event (blank)"?

madmiraal replied:
`to_string()` has a default definition in Object:
https://github.com/godotengine/godot/blob/50db0e66ac8c00f176c706cee0c3a5469453acce/core/object/object.cpp#L800-L809
I'm suggesting not adding the declaration and the dummy definition for `to_string()` to `InputEvent`; only its children.

PS I've tested that `InputEvent` can be made abstract, by making `as_text()` a pure virtual function in `InputEvent`; and removing its dummy definition.

EricEzaM replied:
Ah ok, understood.

Are there any other abstract classes in godot? How do they interact with gd script?

madmiraal replied:
> Are there any other abstract classes in godot? How do they interact with gd script?

Yes. `Shape2D` for example:
https://github.com/godotengine/godot/blob/50db0e66ac8c00f176c706cee0c3a5469453acce/scene/resources/shape_2d.h#L62

madmiraal commented on core/input/input_event.cpp:
```diff
@@ -796,7 +864,26 @@ bool InputEventJoypadMotion::action_match(const Ref<InputEvent> &p_event, bool *
+	TTRC("Left Stick Y-Axis, Joystick 0 Y-Axis"),
+	TTRC("Right Stick X-Axis, Joystick 1 X-Axis"),
+	TTRC("Right Stick X-Axis, Joystick 1 Y-Axis"),
+	TTRC("Joystick 2 X-Axis, Left Trigger, Sony L2, Xbox LT"),
```
Copy and paste error: The fourth entry should be "TTRC("Right Stick **Y**-Axis, Joystick 1 Y-Axis"),"

madmiraal reviewed (approved):
Aside from the typo above, I think it's good to go.

EricEzaM commented:
All sorted now, thanks!

akien-mga commented:
Thanks!

hsandt commented:
Hi, I just got the issue of listening to gamepad axis showing the long readable string, but not the one needed to add an input mapping, in v4.0.beta.custom_build [767f8fb48]

![image](https://user-images.githubusercontent.com/1067772/198884600-bfcdb4ba-507c-46c6-86b4-38cfc0799634.png)

I'm looking at the code in editor/input_event_configuration_dialog.cpp and can clearly see the quickfix under

> // Joypad motion events will display slightly differently than what the event->as_text() provides. See #43660.

but it doesn't seem to be applied.

I couldn't make debugging work in my VS Code so I don't know why the quickfix doesn't work exactly (I'll try to rebuild the whole project with llvm but it will take time, and I have a few things to finish on my project first).

Should I open a new bug report, or let you handle this in this issue?

Zireael07 commented:
This PR has been merged 2 years ago. Open a new issue.
